pytorch_finufft/functional.py

```diff
@@ -836,3 +836,205 @@ def backward(  # type: ignore[override]
+            The resultant array c[j]
+
+        Raises
+        ------
+        ValueError
+            In the case of conflicting specification of the wave-mode ordering.
+        """
+
+        if out is not None:
+            print("In-place results are not yet implemented")
+
+        # TODO -- extend checks to 2d
+        checks._type2_checks(points, targets)
+
+        finufftkwargs = {k: v for k, v in finufftkwargs.items()}
+        _mode_ordering = finufftkwargs.pop("modeord", 1)
+        _i_sign = finufftkwargs.pop("isign", -1)
+
+        if fftshift:
+            if _mode_ordering != 1:  # This seems like it is the wrong way round???????
+                raise ValueError(
+                    "Double specification of ordering; only one of fftshift and "
+                    "modeord should be provided."
+                )
+            _mode_ordering = 0
+
+        ndim = points.shape[0]
+        if _mode_ordering == 1:
+            targets = torch.fft.fftshift(targets, dim=tuple(range(-ndim, 0)))
+
+
+        ctx.isign = _i_sign
+        ctx.mode_ordering = _mode_ordering
+        ctx.fftshift = fftshift
+        ctx.finufftkwargs = finufftkwargs
+
+        ctx.save_for_backward(points, targets)
+
+        nufft_func = get_nufft_func(ndim, 2, points.device.type)
+
+        finufft_out = nufft_func(
+            *points.data.numpy(),
+            targets.data.numpy(),
+            isign=_i_sign,
+            **finufftkwargs,
+        )
+
+        return finufft_out
+
+    @staticmethod
+    def backward(
+        ctx: Any, grad_output: torch.Tensor
+    ) -> Tuple[
+        Union[torch.Tensor, None],
+        Union[torch.Tensor, None],
+        Union[torch.Tensor, None],
+        None,
+        None,
+        None,
+    ]:
+        """
+        Implements derivatives wrt. each argument in the forward method.
+
+        Parameters
+        ----------
+        ctx : Any
+            Pytorch context object
+        grad_output : torch.Tensor
+            Backpass gradient output.
+
+        Returns
+        -------
+        Tuple[ Union[torch.Tensor, None], ...]
+            A tuple of derivatives wrt. each argument in the forward method
+        """
+        _i_sign = ctx.isign
+        _mode_ordering = ctx.mode_ordering
+        finufftkwargs = ctx.finufftkwargs
+
+        points, targets = ctx.saved_tensors
+        device = points.device
+
+        # start_points = -(np.array(targets.shape) // 2)
+        # end_points = start_points + targets.shape
+        # slices = tuple(slice(start, end) for start, end in zip(start_points, end_points))
+
+        # # CPU idiosyncracy that needs to be done differently
+        # k_ramps = torch.from_numpy(np.mgrid[slices], dtype=points.dtype)
+
+        grad_points = grad_targets = None
+
+        ## From type 1, commenting for now to understand whether needed
+        # if any(ctx.needs_input_grad) and _mode_ordering:
+        #     grad_output = torch.fft.fftshift(grad_output)
```

Comment:
Is it surprising that this seems to be not needed?

Reply:
this is a good point. It can't be applied like here because type 2 output is on unstructured points. However, it needs to go somewhere. 

Your comment made me discover that we are not testing different mode ordering in the backwards passes

Reply:
wait no we are ...

Reply:
Are we not? I thought one of the things we are combinatorially testing is fftshift true/false